pkg/controller/workspace/provision/deployment.go

```diff
@@ -0,0 +1,312 @@
+			ProvisioningStatus: ProvisioningStatus{Err: err},
+		}
+	}
+
+	if clusterDeployment == nil {
+		clusterAPI.Logger.Info("Creating deployment...")
+		err := clusterAPI.Client.Create(context.TODO(), specDeployment)
+		return DeploymentProvisioningStatus{
+			ProvisioningStatus: ProvisioningStatus{
+				Requeue: true,
+				Err:     err,
+			},
+		}
+	}
+
+	if !cmp.Equal(specDeployment, clusterDeployment, deploymentDiffOpts) {
+		clusterAPI.Logger.Info("Updating deployment...")
+		clusterDeployment.Spec = specDeployment.Spec
+		err := clusterAPI.Client.Update(context.TODO(), clusterDeployment)
+		if err != nil {
+			if errors.IsConflict(err) {
+				return DeploymentProvisioningStatus{ProvisioningStatus: ProvisioningStatus{Requeue: true}}
+			}
+			return DeploymentProvisioningStatus{ProvisioningStatus{Err: err}}
+		}
+		return DeploymentProvisioningStatus{
+			ProvisioningStatus: ProvisioningStatus{Requeue: true},
+		}
+	}
+
+	deploymentReady := checkDeploymentStatus(clusterDeployment)
+	if deploymentReady {
+		return DeploymentProvisioningStatus{
+			ProvisioningStatus: ProvisioningStatus{
+				Continue: true,
+			},
+		}
+	}
+
+	return DeploymentProvisioningStatus{}
+}
+
+func checkDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
+	return deployment.Status.ReadyReplicas > 0
+}
+
+func getSpecDeployment(
+	workspace *v1alpha1.Workspace,
+	podAdditionsList []v1alpha1.PodAdditions,
+	saName string,
+	scheme *runtime.Scheme) (*appsv1.Deployment, error) {
+	replicas := int32(1)
+	terminationGracePeriod := int64(1)
+	rollingUpdateParam := intstr.FromInt(1)
+
+	var user *int64
+	if !config.ControllerCfg.IsOpenShift() {
+		uID := int64(1234)
+		user = &uID
+	}
+
+	podAdditions, err := mergePodAdditions(podAdditionsList)
+	if err != nil {
+		return nil, err
+	}
+
+	commonEnv := env.CommonEnvironmentVariables(workspace.Name, workspace.Status.WorkspaceId, workspace.Namespace)
+	for idx := range podAdditions.Containers {
+		podAdditions.Containers[idx].Env = append(podAdditions.Containers[idx].Env, commonEnv...)
+	}
+	for idx := range podAdditions.InitContainers {
+		podAdditions.InitContainers[idx].Env = append(podAdditions.InitContainers[idx].Env, commonEnv...)
+	}
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workspace.Status.WorkspaceId,
+			Namespace: workspace.Namespace,
+			Labels: map[string]string{
+				config.WorkspaceIDLabel: workspace.Status.WorkspaceId,
+			},
+			Annotations: map[string]string{
+				config.WorkspaceCreatorAnnotation: workspace.Annotations[config.WorkspaceCreatorAnnotation],
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					config.CheOriginalNameLabel: config.CheOriginalName,
+					config.WorkspaceIDLabel:     workspace.Status.WorkspaceId,
+					config.WorkspaceNameLabel:   workspace.Name,
+				},
+			},
+			Strategy: appsv1.DeploymentStrategy{
+				Type: "RollingUpdate",
```

Comment:
Maybe it's better to start with Recreate since workspaces will fails to update if PVC is used (at least on production-ready cluster)